Make chat exports use ISO 8601/RFC 3339 dates and be more informative

yaya-usman opened the pull request:
Fixes:  [vector-im/element-web#21812](https://github.com/vector-im/element-web/issues/21812)

it also includes the room name in the file export name.

Signed-off-by: Yaya Usman [yaya-usman@users.noreply.github.com](mailto:yaya-usman@users.noreply.github.com) 

**Before:**
![htmlexport](https://user-images.githubusercontent.com/38439166/167714430-680c0deb-e4d5-4b11-bbbc-1fa1fa92fccc.PNG)
![json](https://user-images.githubusercontent.com/38439166/167714467-cdba04f1-4b8d-4674-a419-4cc6285523a2.PNG)
![plaintext](https://user-images.githubusercontent.com/38439166/167714504-81e63ce2-46c2-4e3f-b55a-c52b028bc888.PNG)


**After:** 

![htmlexport](https://user-images.githubusercontent.com/38439166/167713766-62591c50-0b5e-4da2-b572-e1114e107140.PNG)
![json](https://user-images.githubusercontent.com/38439166/167713787-ecddec75-78e8-4daf-8f21-94ea3d170209.PNG)
![plaintext](https://user-images.githubusercontent.com/38439166/167713835-276d4b7f-8667-480f-9296-35e7be959f26.PNG)


<!-- CHANGELOG_PREVIEW_START -->
---
Here's what your changelog entry will look like:

## ✨ Features
 * Make chat exports use ISO 8601/RFC 3339 dates and be more informative ([\#8558](https://github.com/matrix-org/matrix-react-sdk/pull/8558)). Contributed by @yaya-usman.<!-- CHANGELOG_PREVIEW_END -->

turt2live commented on src/utils/exportUtils/JSONExport.ts:
```diff
@@ -108,7 +108,7 @@ export default class JSONExporter extends Exporter {
             await this.downloadZIP();
         } else {
-            const fileName = `matrix-export-${formatFullDateNoDay(new Date())}.json`;
+            const fileName = `${this.room.name}-${formatFullDateNoDayISO(new Date())}.json`;
```
applies throughout: what if the room name contains characters that can't be translated into a file name?

yaya-usman replied:
ohh i understand, i probably should check the room name first, if it contains characters that can't be used as a filename, then i default to the brand name or so, how about that?

turt2live reviewed (changes requested):
Thanks for the PR! While I don't think the Product team cares about the date format as much, they might have thoughts on the file naming otherwise (such as the room name being awkward to look at when made filesafe, or the lack of `brand` variable to denote where the file came from).

Requesting review from the product team to determine if this assumption is true.

kittykat commented:
In what cases is "Element - Chat Export - […]" currently being used? Looking at the before screenshot

yaya-usman commented:
> In what cases is "Element - Chat Export - […]" currently being used? Looking at the before screenshot

When a chat is exported as an HTML

kittykat reviewed (changes requested):
I think the filename should be "`Element-<roomname>-<ISOdate>`" in all cases. For example, `Element-testroom-2022-05-12T12_00_00.735Z.json`

yaya-usman commented:
> I think the filename should be "`Element-<roomname>-<ISOdate>`" in all cases. For example, `Element-testroom-2022-05-12T12_00_00.735Z.json`
> 
> I will leave it to the technical review on what exactly should happen with room names that can't be translated to filenames, but please feel welcome to ask for my input if needed.

Ok thanks, so should i go ahead and implement your suggestion? also concerning the room names that can't be translated to filenames, i already made a suggestion up there like we can check if the room name has some characters that's not acceptable in a filename with regex or so, if it does, we just ignore it and default to the brand name like E`lement-Chat Export-2022-05-12T12_00_00.735Z.json ` what do you think?

kittykat commented:
@yaya-usman Yes, please implement. To confirm, filename should have the following order of preference:

1. `Element-<roomname>-<ISOdate>`
2. If some characters cannot be in the filename, drop those characters and use only ones that can be
3. If that leaves no `<roomname>` component, then drop back to `Element-chat-export-<ISOdate>`

yaya-usman commented:
@kittykat I have made adjustments based on your suggestions. Below images shows the result

**tested room names** 

![testroom](https://user-images.githubusercontent.com/38439166/168403055-b9568586-e121-4f79-a439-f0e88ec61bdc.PNG)

![slashroom](https://user-images.githubusercontent.com/38439166/168403157-b0e915bd-cea5-47fe-a2a4-078971edb325.PNG)

**The following is the resulting filename**
![zip](https://user-images.githubusercontent.com/38439166/168403219-b7a5c2d0-8b5d-444d-80c5-a06ae6aa59dd.PNG)

![zipSlash](https://user-images.githubusercontent.com/38439166/168403182-9ea64665-06ab-4103-80d0-21f64418f081.PNG)



**It works for all the exports**

kittykat reviewed (approved):
LGTM.

One improvement could be to drop any leading and trailing whitespace in room name when the filename is put together. E.g. `Room / /` could give a filename of `Element-Room-2022-05-18T10_28_10.032Z.zip` instead of ``Element-Room  -2022-05-18T10_28_10.032Z.zip`.

yaya-usman commented:
> LGTM.
> 
> One improvement could be to drop any leading and trailing whitespace in room name when the filename is put together. E.g. `Room / /` could give a filename of `Element-Room-2022-05-18T10_28_10.032Z.zip` instead of ``Element-Room -2022-05-18T10_28_10.032Z.zip`.

This means if the room name is say `Empty room` then the filename should be `Element-Emptyroom-2022-05-18T10_28_10.032Z.zip` instead of `Element-Empty room-2022-05-18T10_28_10.032Z.zip` right?

kittykat commented:
> > LGTM.
> > One improvement could be to drop any leading and trailing whitespace in room name when the filename is put together. E.g. `Room / /` could give a filename of `Element-Room-2022-05-18T10_28_10.032Z.zip` instead of ``Element-Room -2022-05-18T10_28_10.032Z.zip`.
> 
> This means if the room name is say `Empty room` then the filename should be `Element-Emptyroom-2022-05-18T10_28_10.032Z.zip` instead of `Element-Empty room-2022-05-18T10_28_10.032Z.zip` right?
develop

Mid-string spaces would stay where they are so:
* `Empty room` to `Element-Empty room-2022-05-18T10_28_10.032Z.zip`
* `Empty room /` to `Element-Empty room-2022-05-18T10_28_10.032Z.zip` as well

t3chguy commented:
This PR contends with https://github.com/matrix-org/matrix-react-sdk/pull/7992

kittykat commented:
@yaya-usman Sorry to follow up again but because of the conflicting PR, can you please split out changing date format into its own PR and tag me in it for product signoff? We can merge that one as soon as possible, and then figure out how to handle the naming conflicts separately :+1: 

Thank you!

yaya-usman commented:
> @yaya-usman Sorry to follow up again but because of the conflicting PR, can you please split out changing date format into its own PR and tag me in it for product signoff? We can merge that one as soon as possible, and then figure out how to handle the naming conflicts separately 👍
> 
> Thank you!

Sure no problem, Sorry for the late response. I will be back fully when I am done with exams in school next weekend.

kittykat commented:
@yaya-usman could you please split out this PR into two? We would like to merge the update to the date format on top of #7992 and then review the other changes to naming separately.

yaya-usman commented:
> @yaya-usman could you please split out this PR into two? We would like to merge the update to the date format on top of #7992 and then review the other changes to naming separately.

Sorry for just responding, i thought this PR has been resolved or rather has been fixed by another person, i deleted the branch from my local but i guess i can still get it back. Also how do i split the PR, i don't really understand that part

kittykat commented:
> > @yaya-usman could you please split out this PR into two? We would like to merge the update to the date format on top of #7992 and then review the other changes to naming separately.
> 
> Sorry for the late response, i thought this PR has been resolved or rather has been fixed by another person, i deleted the branch from my local but i guess i can still get it back. Also how do i split the PR, i don't really understand that part

Here's a guide you could have a look at [for splitting a PR](https://gist.github.com/loilo/930f141d9acf89e9e734ffa042acd750) and to split a commit, I would use [an interactive rebase, edit the commit](http://shafiul.github.io/gitbook/4_interactive_rebasing.html), then do [an interactive add](https://git-scm.com/search/results?search=git%20add)

turt2live commented:
Thanks for the PR! After asking for things to be split up we talked about it more as a group and came to a slightly different conclusion on the file naming. That conclusion is now documented as https://github.com/matrix-org/matrix-react-sdk/pull/9440 (which includes your other PR in it). Hopefully this means the code lands sooner - apologies for the delay here.
